Версия 4. <feat> Добавлен класс ErrorHandler для обработки исключений.

Solntmore opened the pull request:
<feat> Добавлен класс FilmIdCounter для присвоения id фильмам.
          <feat> Добавлен класс ErrorResponse, создающий шаблон ответа по ошибке.
          <feat> Добавлены классы FilmService и UserService, реализующие бизнес-логику приложения.
          <feat> Добавлены классы InMemoryFilmStorage и InMemoryUserStorage реализующие процесс хранения объектов.
          <feat> Добавлены интерфейсы FilmStorage и UserStorage.

Solntmore committed:
Версия 4. <feat> Добавлен класс ErrorHandler для обработки исключений.
          <feat> Добавлен класс FilmIdCounter для присвоения id фильмам.
          <feat> Добавлен класс ErrorResponse, создающий шаблон ответа по ошибке.
          <feat> Добавлены классы FilmService и UserService, реализующие бизнес-логику приложения.
          <feat> Добавлены классы InMemoryFilmStorage и InMemoryUserStorage реализующие процесс хранения объектов.
          <feat> Добавлены интерфейсы FilmStorage и UserStorage.

Solntmore committed:
Версия 4. <feat> Добавлен класс ErrorHandler для обработки исключений.
          <feat> Добавлен класс FilmIdCounter для присвоения id фильмам.
          <feat> Добавлен класс ErrorResponse, создающий шаблон ответа по ошибке.
          <feat> Добавлены классы FilmService и UserService, реализующие бизнес-логику приложения.
          <feat> Добавлены классы InMemoryFilmStorage и InMemoryUserStorage реализующие процесс хранения объектов.
          <feat> Добавлены интерфейсы FilmStorage и UserStorage.
          <refactor> Адаптированы тестовые классы FilmControllerTest и UserControllerTest.

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,56 @@
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public ErrorResponse handleSystemExceptions(final NullPointerException e) {
+        return new ErrorResponse(e.getMessage());
+    }
```
в целом этот хендлер не очень нужен, т.к. при NPE приложение все равно отдаст 500 ошибку, не зависимо отловится она хэндлером или нет

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,55 +2,69 @@
-        log.debug("Получен запрос GET /films. Текущее количество фильмов: {}", films.size());
-        return films.values();
+    public Collection<Film> findAllFilms() {
+        return inMemoryFilmStorage.findAllFilms();
```
контроллер не должен напрямую взаимодейстовать с классами стореджа/репозитория, только через соответствующие методы сервиса 

можно сказать взаимодействие со стераджами должно быть "защищено" прослойкой в видео сервисов

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,64 +1,69 @@
+    }
+
+    @GetMapping("/{userId}/friends/common/{friendId}")
+    public ArrayList<User> getCommonFriends(@PathVariable int userId, @PathVariable int friendId) {
```
лучше возвращать более широкий тип List<User>, так полиморфизм будет работать в полной мере, 
напимер в будущем реализация showCommonFriends может поменятся и он будет возвращать LinckedList, с текущей вариантом, это будет невозможно

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,13 +1,16 @@
-
+import java.util.Set;
+import java.util.TreeSet;
+@Slf4j
```
должна быть пустая строка между аннтоциями и импортами

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -17,6 +20,7 @@ public class User {
     private String login;
     private String name;
     private LocalDate birthday;
+    private ArrayList<Integer> friends = new ArrayList<>();
```
похожий момент с комментарием в ЮзерКонтроллере,
здесь тип переменной лучше сделать более широким List

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+
+    public ArrayList<Film> showMostLikedFilms(int count) {
```
аналагично с более широким возвращаемым типом List

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -0,0 +1,66 @@
+        if (k > count)
+            films.subList(count, k).clear();
+        return films;
+    }
```
неплохая рабочая логика, но попробуй ее сделать через stream, 
мне кажется, должно получится поаккуратней

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -0,0 +1,92 @@
+
+    public void addToFriends(int userId, int friendId) {
+        if (inMemoryUserStorage.findUserById(userId).isPresent()
+                && inMemoryUserStorage.findUserById(friendId).isPresent()) {
```
inMemoryUserStorage.findUserById(...) - можно вынести в локальные переменные, тогда не нужно будет несолько раз доставать один и тот же объет, 
+ код будет восприниматься чуть легче

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -0,0 +1,92 @@
+                throw new InvalidIdException("Пользователи уже являются друзьями, попробуйте другой id.");
+            }
+        } else {
+            log.debug("Неверно введен id {}.");
```
в целом, этот уровень может выше либо info, лиюо даже warn

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -0,0 +1,92 @@
+            log.debug("Неверно введен id {}.");
+            throw new InvalidIdException("Неверно введен id пользователя или друга.");
+        }
+    }
```
здесь аналагично с пред методом

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -0,0 +1,92 @@
+        } else {
+            throw new InvalidIdException("Неверно введен id пользователя или друга.");
+        }
+    }
```
мне кажется эот метод тоже можно попробовать частично или полностью сделать через stream

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -0,0 +1,92 @@
+        }
+        throw new InvalidIdException("Неверно введен id пользователя.");
+    }
+}
```
и этот ))

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,66 @@
+public class InMemoryUserStorage implements UserStorage {
+    private final Map<Integer, User> users = new HashMap<>();
+    private final UserValidator userValidator = new UserValidator();
+    private final UserIdCounter idCounter = new UserIdCounter();
```
здесь аналагично с FilmStorage

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,58 @@
+    public Film createFilm(@Valid Film film) {
+        log.debug("Получен запрос POST /films.");
+        filmValidator.validator(film);
+        film.setId(idCounter.incrementIdCounter());
```
на самом деле, 
можно не создавать отдельные классы UserIdCounter и FilmIdCounter, а методы которые внутри написаны, просто перенести в  InMemory.... классы

т.к. в InMemory классах должна содержаться вся необходимая логика взаимодействия со стореджом, в т.ч. и айдикаунтер

В будущем, когда добавиться реализация с БД, айди будет автоматически инкрементится в рамках самой бд. и IdCounter классы будут не нужны

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,58 @@
+@Component
+public class InMemoryFilmStorage implements FilmStorage {
+    private final Map<Integer, Film> films = new HashMap<>();
+    private final FilmValidator filmValidator = new FilmValidator();
```
валидатор можно было бы сделать тоже через спринг компонент, а потом его заавтовайрить

Solntmore committed:
Версия 5. <refactor> Контроллеры получают доступ к методам storage классов через прослойку в виде UserService и FilmService.
          <refactor> В классах UserService и FilmService переработана логика работы с данными, добавлено использование стримов.
          <refactor> Прочие мелкие исправления.

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -52,15 +55,25 @@ public void removeLikeFromFilm(int filmId, int userId) {
+        return inMemoryFilmStorage.findAllFilms()
+                .stream().sorted(Comparator.comparingLong(Film::showQuantityOfLikes).reversed())
+                .limit(count).collect(Collectors.toList());
+    }
```
супер, это совсем другой уровень)

Fleaman777 commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -22,71 +23,71 @@ public UserService(InMemoryUserStorage inMemoryUserStorage) {
+                            .orElseThrow(() -> new InvalidIdException("Нет друга с таким id"))
+                            .getFriends().contains(id))
+                    .map(commonId -> inMemoryUserStorage.findUserById(commonId).orElse(null))
+                    .collect(Collectors.toList());
```
высший пилотаж) очень хорошо выглядит
